File: leetcode/practice/576-out-of-boundary-paths/dp.java
class Solution
{
    int[][][] dp ;
    int mod = 1000000007 ;
    public int solve(int m , int n , int maxMove , int r , int c)
    {
        if(r < 0 || r >= m || c < 0 || c >= n)
            return 1 ;
        if(maxMove <= 0)
            return 0 ;

        if(dp[r][c][maxMove] != -1)
            return dp[r][c][maxMove] ;
        int left = solve(m , n , maxMove - 1 , r , c - 1)%mod ;
        int right = solve(m , n , maxMove - 1 , r , c + 1)%mod ;
        int up = solve(m , n , maxMove - 1 , r - 1 , c)%mod ;
        int down = solve(m , n , maxMove - 1 , r + 1 , c)%mod ;
        dp[r][c][maxMove] = ((left + right)%mod + (up + down)%mod)%mod ; 
        return dp[r][c][maxMove] ;
    }

    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn)
    {
        dp = new int[m][n][maxMove + 1] ;
        for(int i = 0 ; i < m ; i++)
            for(int j = 0 ; j < n ; j++)
                for(int k = 0 ; k <= maxMove ; k++)
                    dp[i][j][k] = -1 ;

        return solve(m , n , maxMove , startRow , startColumn) ;
    }
}

File: leetcode/weekly-contests/contest-201/1547-minimum-cost-to-cut-a-stick/dp.cpp
struct pair_hash
{
    template <class T1, class T2>
    std::size_t operator() (const std::pair<T1, T2> &pair) const {
        return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
    }
};
class Solution
{
    unordered_map<pair<int , int> , int , pair_hash>dp ;
public:
    int solve(vector<int>& cuts , int left , int right)
    {
        if(right - left <= 1)
            return 0 ;
        if(dp.find({left , right}) != dp.end())return dp[{left , right}] ;

        int minValue = INT_MAX ;

        for(auto ele : cuts)
        {
            if(ele >= right or ele <= left)continue ;

            int temp = 0 ;
            temp += solve(cuts , left , ele) ;
            temp += solve(cuts , ele , right) ;

            // cost of cutting left and right part plus the
            // cost of cutting current part .
            minValue = min(minValue , temp + right - left) ;
        }

        if(minValue == INT_MAX)
            minValue = 0 ;
        dp[{left , right}] = minValue ;
        return minValue ;
    }
    int minCost(int n, vector<int>& cuts)
    {
        return solve(cuts , 0 , n) ;
    }
};
